luarules/gadgets/game_unit_sharing_mode.lua

```diff
@@ -0,0 +1,101 @@
+local gadget = gadget ---@type Gadget
+
+function gadget:GetInfo()
+	return {
+		name    = 'Unit Sharing Control',
+		desc    = 'Controls unit sharing based on modoption settings',
+		author  = 'Rimilel',
+		date    = 'May 2024 / April 2025',
+		license = 'GNU GPL, v2 or later',
+		layer   = 0,
+		enabled = true
+	}
+end
+
+----------------------------------------------------------------
+-- Synced only
+----------------------------------------------------------------
+if not gadgetHandler:IsSyncedCode() then
+	return false
+end
+
+local unitSharingMode = Spring.GetModOptions().unit_sharing_mode or "enabled"
+local unitMarketEnabled = Spring.GetModOptions().unit_market or false
+
+-- Disable the gadget only if unit sharing is fully enabled
+if unitSharingMode == "enabled" then
+	return false
+end
+
+--[[ Engine Callin Note: AllowUnitTransfer blocks if *any* gadget returns false.
```

Comment:
Thank you for clarifying this, might also want to add that returning `true` does not ensure the unit can be shared, just that it allows other gadgets to check if they want to allow it to be shared.